src/engine/OrderBy.cpp

```diff
@@ -77,6 +77,19 @@ void OrderBy::computeResult(ResultTable* result) {
   AD_CHECK(!_sortIndices.empty());
   shared_ptr<const ResultTable> subRes = _subtree->getResult();
 
+  // TODO<joka921> proper timeout and better estimate
+  double remainingSecs =
+      static_cast<double>(_timeoutTimer->wlock()->remainingMicroseconds()) /
+      1000000;
+  if (static_cast<double>(subRes->size()) / 50000000 > remainingSecs) {
+    // assume that we can sort at most 50M elements per second. This is a rather
+    // generous upper bound to make sure we do not exceed our timeout more than
+    // an order of magnitude b.c of sorts
```

Comment:
In general: You are right with all the three points. My approach to this was using the simplified approach (sort is linear, and the payload can be neglected). I used a rather large sort with a relatively high number of columns, and doubled its time and came up with this constant. This way, we compensate for the two inaccuracies. My goal was to underestimate the sort time / overestimate the sort speed, such that we would never cancel a sort that could be computed. 

In detail:

1. Of course we can do this. We can even sort a variety of inputs (e.g. 3 different sizes with 1, 5, 15 columns) and fit a sort curve on this. The question is, whether this is worth it at this point. For the autocompletion experiments on our machine, this approach was sufficient, but it will break easily if we have much faster machines (more Cores for sorting at the same time) or much smaller timeout values (much less than a second).

2. As we all know it is not linear, but close enough to linear in practice.
3. From my experience, this is a real issue, the number of payload columns is much more relevant than the small logarithmic term in the theoretical complexity. However, linear in the number of columns would be an overestimation (the payload has to be copied, which is still easier than doing additional comparisons and branch based on their results, which has to be done for additional inputs.